### PR TITLE
Optimizations for multiple quadword precision integer support (multi-quadword).

### PR DESCRIPTION
**Pull request by munroesj52**
Added multi-quadword multiply-add operations optimized for (POWER7/8/9).
Used them to optimize larger multi-quadword multiply while carefully
use COMPILER_FENCE to avoid register presure and spill code.
Added appropriate unit, compile, anf performance tests.

	* src/pveclib/vec_int512_ppc.h General doxygen text cleanup.
	[i512_security_issues_0_0_2]: Remove todos from subsection.
	[i512_Endian_issues_0_0]: Remove todos from section.
	(__VEC_U_512x1): Add x2 field to union in support of madd.
	(__VEC_U_4096x512): Add x4 field to union in support of add512.
	[__VEC_EXPLICITE_FENCE_NOPS__]: Enable explicite NOPS to make
	compiler fences visible.
	(vec_mul256x256_inline): Add doxygen note.
	Leverage madduq/madd2uq.
	(vec_mul512x128_inline): Add doxygen note. Leverage madduq.
	(vec_madd512x128a128_inline, ec_madd512x128a512_inline,
	vec_madd512x128a128a512_inline): New operation.
	(vec_mul512x512_inline): Add doxygen note.
	Leverage vec_madd512x128a512_inline.
	(vec_madd512x512a512_inline): New operation.
	(vec_mul128x128, vec_mul256x256): Update latency table.
	(vec_mul1024x1024, vec_mul2048x2048):
	Add note and update latency table.
	(vec_mul128_byMN, vec_mul512_byMN) New library functions.
	(vec_madd512x128a128, vec_madd512x128a512,
	vec_madd512x128a128a512, vec_madd512x512a512,
	vec_mul128_byMN, vec_mul512_byMN):
	New platform qualified externs.

	src/vec_int512_runtime.c [__VEC_EXPLICITE_FENCE_NOPS__]:
	Define COMPILE_FENCE, COMPILE_FENCE1, COMPILE_FENCE2,
	COMPILE_FENCE3, COMPILE_FENCE10, COMPILE_FENCE11,
	COMPILE_FENCE12, COMPILE_FENCE13, COMPILE_FENCE14,
	COMPILE_FENCE15, COMPILE_FENCE16, COMPILE_FENCE17,
	COMPILE_FENCE20, COMPILE_FENCE21, COMPILE_FENCE22,
	COMPILE_FENCE23, COMPILE_FENCE24, COMPILE_FENCE25,
	COMPILE_FENCE26, COMPILE_FENCE27, COMPILE_FENCE28,
	COMPILE_FENCE29):
	(vec_mul256x256, vec_mul512x128): Optimized for static
	implementation for PWR8/9.
	(vec_madd512x128a128, vec_madd512x128a512,
	vec_madd512x128a128a512): New function.
	(vec_mul512x512): Optimized for static
	implementation for PWR8/9.
	(vec_madd512x512a512):  New function.
	(vec_mul1024x1024 [__VEC_USE_RESTRICT__]):
	Disable restrict for now.
	(vec_mul1024x1024):
	Inline optimized PWR8/9 512-bit mul/madd implementions.
	(vec_mul2048x2048 [__VEC_USE_RESTRICT__]):
	Disable restrict for now.
	(vec_mul2048x2048):
	Inline optimized PWR8/9 512-bit mul/madd implementions.
	(vec_mul128_byMN, vec_mul512_byMN): New library functions.

	* src/vec_runtime_DYN.c ([__ORDER_BIG_ENDIAN__]
	vec_mul128_byMN_PWR7, vec_mul512_byMN_PWR7): New externs.
	([__ORDER_LITTLE_ENDIAN__] vec_mul128_byMN_PWR9,
	vec_mul512_byMN_PWR9): New externs.
	(vec_mul128_byMN_PWR8, vec_mul512_byMN_PWR8): New externs.
	(resolve_vec_mul128_byMN, resolve_vec_mul512_byMN):
	New resolver.
	(vec_mul128_byMN, vec_mul512_byMN): New IFUNC symbol.

	* src/testsuite/arith128_test_i512.c
	(vec512_ten2048_13, vec512_ten2048_12, vec512_ten2048_11,
	vec512_ten2048_10. vec512_ten2048_9, vec512_ten2048_8,
	vec512_ten2048_7, vec512_ten2048_6, vec512_ten2048_5,
	vec512_ten2048_4, vec512_ten2048_3, vec512_ten2048_2,
	vec512_ten2048_1, vec512_ten2048_0): Static const for 10^2048.
	(test_time_i512): Add timed Loops for timed_mul128x128,
	timed_mul256x256. timed_mul512x512by8, timed_mul1024x1024by8,
	timed_mul2048x2048by8, timed_mul2048x2048_MN, and
	timed_mul4096x4096_MN.
	(test_mul512x128_MN, test_madd512x128, test_mul512x512_MN,
	test_mul2048x2048_MN): New unit tests.
	(test_vec_i512): Add calls to unit tests test_madd512x128,
	test_mul512x128_MN, test_mul512x512_MN, and
	test_mul2048x2048_MN.
	* src/testsuite/arith128_test_i512.h
	(vec512_ten2048_13, vec512_ten2048_12, vec512_ten2048_11,
	vec512_ten2048_10. vec512_ten2048_9, vec512_ten2048_8,
	vec512_ten2048_7, vec512_ten2048_6, vec512_ten2048_5,
	vec512_ten2048_4, vec512_ten2048_3, vec512_ten2048_2,
	vec512_ten2048_1, vec512_ten2048_0): Add const externs.
	(test_mul1024x1024): Add function extern.

	* src/testsuite/vec_perf_i512.c (c_zero, c_one, c_ten,
	c_hundred, c_10k, c_100m, ten_64h,  ten_64l):
	New static consts.
	(timed_mul128x128, timed_mul256x256. timed_mul512x512by8,
	timed_mul1024x1024by8, timed_mul2048x2048by8,
	timed_mul2048x2048_MN, timed_mul4096x4096_MN):
	New timed functions.
	* src/testsuite/vec_perf_i512.h (timed_mul128x128,
	timed_mul256x256, timed_mul512x512by8, timed_mul1024x1024by8,
	timed_mul2048x2048by8, timed_mul2048x2048_MN,
	timed_mul4096x4096_MN): Externs for timed functions.

Signed-off-by: Steven Munroe <munroesj52@gmail.com>

### CODE REVIEW

File: src/vec_int512_runtime.c [hunk 1]
```diff
@@ -31,19 +80,256 @@ __VEC_PWR_IMP (vec_mul128x128) (vui128_t m1l, vui128_t m2l)
...
+
+  result.vx0 = mpll;
+  result.vx1 = mplh;
+  result.vx2 = mphl;
+  result.vx3 = mphh;
+  return result;
+#else
+#ifdef _ARCH_PWR8
+  __VEC_U_512 result;
+  vui128_t mc, mp, mq, mqhl;
```
Comment (ThinkOpenly) [inline src/vec_int512_runtime.c:109]: mc unused?

File: src/vec_int512_runtime.c [hunk 2]
```diff
@@ -31,19 +80,256 @@ __VEC_PWR_IMP (vec_mul128x128) (vui128_t m1l, vui128_t m2l)
...
+  result.vx0 = mpx0;
+  result.vx1 = mpx1;
+  result.vx2 = mpx2;
+  result.vx3 = mpx3;
+  result.vx4 = mq3;
+  return result;
+#else
+#ifdef _ARCH_PWR8
+  __VEC_U_640 result;
+  vui128_t mq3, mq2, mq1, mq0, mq, mc;
```
Comment (ThinkOpenly) [inline src/vec_int512_runtime.c:161]: mq and mc unused.

File: src/vec_int512_runtime.c [hunk 3]
```diff
@@ -31,19 +80,256 @@ __VEC_PWR_IMP (vec_mul128x128) (vui128_t m1l, vui128_t m2l)
...
+  return vec_madd512x128a128_inline (m1, m2, a1);
+}
+
+__VEC_U_640
+__VEC_PWR_IMP (vec_madd512x128a512) (__VEC_U_512 m1, vui128_t m2,
+                                     __VEC_U_512 a2)
+{
+#ifdef _ARCH_PWR9
+  __VEC_U_640 result;
+  vui128_t mq3, mq2, mq1, mq0, mq, mc;
```
Comment (ThinkOpenly) [inline src/vec_int512_runtime.c:194]: mq, mc unused.

File: src/vec_int512_runtime.c [hunk 4]
```diff
@@ -31,19 +80,256 @@ __VEC_PWR_IMP (vec_mul128x128) (vui128_t m1l, vui128_t m2l)
...
+  result.vx0 = mpx0;
+  result.vx1 = mpx1;
+  result.vx2 = mpx2;
+  result.vx3 = mpx3;
+  result.vx4 = mq3;
+  return result;
+#else
+#ifdef _ARCH_PWR8
+  __VEC_U_640 result;
+  vui128_t mq3, mq2, mq1, mq0, mq, mc;
```
Comment (ThinkOpenly) [inline src/vec_int512_runtime.c:214]: mq, mc unused.

File: src/vec_int512_runtime.c [hunk 5]
```diff
@@ -31,19 +80,256 @@ __VEC_PWR_IMP (vec_mul128x128) (vui128_t m1l, vui128_t m2l)
...
+  return vec_madd512x128a128a512_inline (m1, m2, a1, a2);
 }
 
 __VEC_U_1024 __attribute__((flatten))
 __VEC_PWR_IMP (vec_mul512x512) (__VEC_U_512 m1, __VEC_U_512 m2)
 {
+#ifdef _ARCH_PWR8
+  __VEC_U_1024 result;
+#ifdef _ARCH_PWR9
+  vui128_t mc, mp, mq;
```
Comment (ThinkOpenly) [inline src/vec_int512_runtime.c:249]: mc unused

File: src/vec_int512_runtime.c [hunk 6]
```diff
@@ -31,19 +80,256 @@ __VEC_PWR_IMP (vec_mul128x128) (vui128_t m1l, vui128_t m2l)
...
+  COMPILE_FENCE12;
+
+  mp3 = __VEC_PWR_IMP (vec_mul512x128) (m1, m2.vx3);
+  result.vx3 = vec_addcq (&mq, mp3.vx0, result.vx3);
+  result.vx4 = vec_addeq (&mp, mp3.vx1, result.vx4, mq);
+  result.vx5 = vec_addeq (&mq, mp3.vx2, result.vx5, mp);
+  result.vx6 = vec_addeq (&mp, mp3.vx3, result.vx6, mq);
+  result.vx7 = vec_addeuqm (result.vx7, mp3.vx4, mp);
+#else
+  vui128_t mc, mp, mq;
```
Comment (ThinkOpenly) [inline src/vec_int512_runtime.c:280]: all unused

File: src/vec_int512_runtime.c [hunk 7]
```diff
@@ -31,19 +80,256 @@ __VEC_PWR_IMP (vec_mul128x128) (vui128_t m1l, vui128_t m2l)
...
+}
+
+__VEC_U_1024 __attribute__((flatten))
+__VEC_PWR_IMP (vec_madd512x512a512) (__VEC_U_512 m1, __VEC_U_512 m2,
+                                     __VEC_U_512 a1)
+{
+#ifdef _ARCH_PWR8
+  __VEC_U_1024 result;
+  __VEC_U_512x1 mp3, mp2, mp1, mp0;
+  vui128_t mc, mp, mq;
```
Comment (ThinkOpenly) [inline src/vec_int512_runtime.c:312]: all unused

File: src/vec_int512_runtime.c [hunk 8]
```diff
@@ -136,103 +454,370 @@ __VEC_PWR_IMP (vec_mul2048x2048) (__VEC_U_4096* r4096,
...
+// Adjust the order of quadwords in multiple quadword precision value,
+// to match the endian order of the bytes in a quadword.
+#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+#define __MDX(__index) (__index)
+#define __NDX(__index) (__index)
+#define __PDX(__index) (__index)
+#else
+#define __NDX(__index) ((nx - 1) - (__index))
+#define __MDX(__index) ((mx - 1) - (__index))
+#define __PDX(__index) ((px - 1) - (__index))
```
Comment (ThinkOpenly) [inline src/vec_int512_runtime.c:669]: put these in the same order?

File: src/pveclib/vec_int512_ppc.h [hunk 9]
```diff
@@ -1619,10 +1646,11 @@ typedef union
  *  and code motion to smaller code blocks. This in turn reduces
  *  register pressure and avoids generating spill code.
  */
-#if 1
-#define COMPILE_FENCE __asm (";":::)
-#else
+#ifdef __VEC_EXPLICITE_FENCE_NOPS__
```
[the remark this answers is not in the document]
Comment (munroesj52) [reply src/pveclib/vec_int512_ppc.h:1649]: Yes ;)

File: src/pveclib/vec_int512_ppc.h [hunk 10]
```diff
@@ -1909,6 +1937,9 @@ vec_mul256x256_inline (__VEC_U_256 m1, __VEC_U_256 m2)
  *  The product is returned as single 640-bit integer in a
  *  homogeneous aggregate structure.
  *
+ *  \note Using the Multiple-Add form which applies the addend early
```
[the remark this answers is not in the document]
Comment (munroesj52) [reply src/pveclib/vec_int512_ppc.h:1940]: OK

File: src/pveclib/vec_int512_ppc.h [hunk 11]
```diff
@@ -1861,6 +1889,9 @@ vec_mul128x128_inline (vui128_t a, vui128_t b)
  *  The product is returned as single 512-bit integer in a
  *  homogeneous aggregate structure.
  *
+ *  \note Using the Multiple-Add form which applies the addend early
```
[the remark this answers is not in the document]
Comment (munroesj52) [reply src/pveclib/vec_int512_ppc.h:1892]: OK

File: src/pveclib/vec_int512_ppc.h [hunk 12]
```diff
@@ -1929,18 +1960,160 @@ vec_mul512x128_inline (__VEC_U_512 m1, vui128_t m2)
...
+ *  |--------:|:-----:|:---------|
+ *  |power8   |224-232| 1/cycle  |
+ *  |power9   |132-135| 1/cycle  |
+ *
+ *  @param m1 vector representation of a unsigned 512-bit integer.
+ *  @param m2 vector representation of a unsigned 128-bit integer.
+ *  @param a1 vector representation of a unsigned 128-bit integer.
+ *  @param a2 vector representation of a unsigned 512-bit integer.
+ *  @return homogeneous aggregate representation of the unsigned
+ *  640-bit sum of (m1 * m2) + a1.
```
[the remark this answers is not in the document]
Comment (munroesj52) [reply src/pveclib/vec_int512_ppc.h:2103]: OK

File: src/pveclib/vec_int512_ppc.h [hunk 13]
```diff
@@ -1929,18 +1960,160 @@ vec_mul512x128_inline (__VEC_U_512 m1, vui128_t m2)
   __VEC_U_640 result;
   vui128_t mq3, mq2, mq1, mq0, mq, mc;
```
[the remark this answers is not in the document]
Comment (munroesj52) [reply src/pveclib/vec_int512_ppc.h:1961]: Ok

File: src/pveclib/vec_int512_ppc.h [hunk 14]
```diff
@@ -1929,18 +1960,160 @@ vec_mul512x128_inline (__VEC_U_512 m1, vui128_t m2)
...
+ *  @param m2 vector representation of a unsigned 128-bit integer.
+ *  @param a1 vector representation of a unsigned 128-bit integer.
+ *  @return homogeneous aggregate representation of the unsigned
+ *  640-bit sum of (m1 * m2) + c.
+ */
+static inline __VEC_U_640
+vec_madd512x128a128_inline (__VEC_U_512 m1, vui128_t m2, vui128_t a1)
+{
+  __VEC_U_640 result;
+  vui128_t mq3, mq2, mq1, mq0, mq, mc;
```
[the remark this answers is not in the document]
Comment (munroesj52) [reply src/pveclib/vec_int512_ppc.h:2010]: ok

File: src/pveclib/vec_int512_ppc.h [hunk 15]
```diff
@@ -1929,18 +1960,160 @@ vec_mul512x128_inline (__VEC_U_512 m1, vui128_t m2)
...
+ *  @param m2 vector representation of a unsigned 128-bit integer.
+ *  @param a2 vector representation of a unsigned 512-bit integer.
+ *  @return homogeneous aggregate representation of the unsigned
+ *  640-bit sum of (m1 * m2) + a1.
+ */
+static inline __VEC_U_640
+vec_madd512x128a512_inline (__VEC_U_512 m1, vui128_t m2, __VEC_U_512 a2)
+{
+  __VEC_U_640 result;
+  vui128_t mq3, mq2, mq1, mq0, mq, mc;
```
[the remark this answers is not in the document]
Comment (munroesj52) [reply src/pveclib/vec_int512_ppc.h:2059]: ok

File: src/pveclib/vec_int512_ppc.h [hunk 16]
```diff
@@ -1929,18 +1960,160 @@ vec_mul512x128_inline (__VEC_U_512 m1, vui128_t m2)
...
+ *  @param a1 vector representation of a unsigned 128-bit integer.
+ *  @param a2 vector representation of a unsigned 512-bit integer.
+ *  @return homogeneous aggregate representation of the unsigned
+ *  640-bit sum of (m1 * m2) + a1.
+ */
+static inline __VEC_U_640
+vec_madd512x128a128a512_inline (__VEC_U_512 m1, vui128_t m2, vui128_t a1, __VEC_U_512 a2)
+{
+  __VEC_U_640 result;
+  vui128_t mq3, mq2, mq1, mq0, mq, mc;
```
[the remark this answers is not in the document]
Comment (munroesj52) [reply src/pveclib/vec_int512_ppc.h:2109]: ok

File: src/pveclib/vec_int512_ppc.h [hunk 17]
```diff
@@ -1952,14 +2125,18 @@ vec_mul512x128_inline (__VEC_U_512 m1, vui128_t m2)
...
- *  Compute the 640 bit product of 512 bit values m1 and m2.
+ *  Compute the 1024 bit product of 512 bit values m1 and m2.
  *  The product is returned as single 1024-bit integer in a
  *  homogeneous aggregate structure.
  *
  *  \note We use the COMPILER_FENCE to limit instruction scheduling
  *  and code motion to smaller code blocks. This in turn reduces
  *  register pressure and avoids generating spill code.
  *
+ *  \note Using the Multiple-Add form which applies the addend early
```
[the remark this answers is not in the document]
Comment (munroesj52) [reply src/pveclib/vec_int512_ppc.h:2136]: ok

File: src/pveclib/vec_int512_ppc.h [hunk 18]
```diff
@@ -1975,32 +2152,77 @@ vec_mul512x512_inline (__VEC_U_512 m1, __VEC_U_512 m2)
 {
   __VEC_U_1024 result;
   vui128_t mc, mp, mq;
```
[the remark this answers is not in the document]
Comment (munroesj52) [reply src/pveclib/vec_int512_ppc.h:2154]: ok

File: src/pveclib/vec_int512_ppc.h [hunk 19]
```diff
@@ -1975,32 +2152,77 @@ vec_mul512x512_inline (__VEC_U_512 m1, __VEC_U_512 m2)
...
+ *  live ranges for registers passing partial products for larger
+ *  multiple precision multiplies.
+ *  \note We use the COMPILER_FENCE to limit instruction scheduling
+ *  and code motion to smaller code blocks. This in turn reduces
+ *  register pressure and avoids generating spill code.
+ *
+ *  |processor|Latency|Throughput|
+ *  |--------:|:-----:|:---------|
+ *  |power8   | ?? | 1/cycle  |
+ *  |power9   | ?? | 1/cycle  |
```
[the remark this answers is not in the document]
Comment (munroesj52) [reply src/pveclib/vec_int512_ppc.h:2195]: Oversight. Copied timing from library vec_mul512x512.

File: src/pveclib/vec_int512_ppc.h [hunk 20]
```diff
@@ -1975,32 +2152,77 @@ vec_mul512x512_inline (__VEC_U_512 m1, __VEC_U_512 m2)
...
+ *  @param m2 vector representation of a unsigned 512-bit integer.
+ *  @param a1 vector representation of a unsigned 512-bit integer.
+ *  @return homogeneous aggregate representation of the unsigned
+ *  1028-bit product of a * b.
+ */
+static inline __VEC_U_1024
+vec_madd512x512a512_inline (__VEC_U_512 m1, __VEC_U_512 m2, __VEC_U_512 a1)
+{
+  __VEC_U_1024 result;
+  vui128_t mc, mp, mq;
```
[the remark this answers is not in the document]
Comment (munroesj52) [reply src/pveclib/vec_int512_ppc.h:2207]: OK

File: src/pveclib/vec_int512_ppc.h [hunk 21]
```diff
@@ -2148,6 +2385,72 @@ extern void
...
 
+/** \brief Vector Unsigned Integer Quadword MxN Multiply.
+ *
+ *  Compute the M+N quadword product of two quadword arrays  m1, m2.
+ *  The product is returned as M+N quadword array p.
+ *
+ *  \note This is the dynamic call ABI for IFUNC selection.
+ *  The static implementations are vec_mul128_byMN_PWR8 and
+ *  vec_mul128_byMN_PWR9. For static calls the __VEC_PWR_IMP() macro
+ *  will add appropriate suffix based on the compile -mcpu= option.
```
[the remark this answers is not in the document]
Comment (munroesj52) [reply src/pveclib/vec_int512_ppc.h:2396]: yes really. There more work to clean up the config/build as what I have is not elegant, but this works.

File: src/pveclib/vec_int512_ppc.h [hunk 22]
```diff
@@ -2148,6 +2385,72 @@ extern void
...
+ *  quadword element 0. The most significant is quadword elements
+ *  [M-1], [N-1], and [M+N-1].
+ *  On Big Endian systems the least significant quadword is
+ *  quadword elements [M-1], [N-1], and [M+N-1].
+ *  The most significant is quadword element 0.
+ *
+ *  |processor|Latency|Throughput|
+ *  |--------:|:-----:|:---------|
+ *  |power8   | ??? | 1/cycle  |
+ *  |power9   | ??? | 1/cycle  |
```
[the remark this answers is not in the document]
Comment (munroesj52) [reply src/pveclib/vec_int512_ppc.h:2408]: OK, I just have the measured data for this yet.

File: src/testsuite/vec_perf_i512.c [hunk 23]
```diff
@@ -621,6 +887,108 @@ timed_mul2048x2048 (void)
   return (rc);
 }
 
+  /* Older (GCC6) compilers ignore the "no-unroll-loops" pragma and make
+   it hard to isolate each iteration of vec_muludq.  This version
+   calls __test_muludq from vec_int128_dummy.c as a work around.  */
```
[the remark this answers is not in the document]
Comment (munroesj52) [reply src/testsuite/vec_perf_i512.c:892]: Out of context pasto. Removed. In the int512 case always calling out-of-line so also removed the redundant  __attribute__((optimize ("unroll-loops")))

File: src/vec_int512_runtime.c (hunk 1 above)
Comment (munroesj52) [reply src/vec_int512_runtime.c:109]: Removed mc and mqhl

File: src/vec_int512_runtime.c (hunk 2 above)
Comment (munroesj52) [reply src/vec_int512_runtime.c:161]: Removed , mq, mc,

File: src/vec_int512_runtime.c (hunk 3 above)
Comment (munroesj52) [reply src/vec_int512_runtime.c:194]: Removed , mq, mc

File: src/vec_int512_runtime.c (hunk 4 above)
Comment (munroesj52) [reply src/vec_int512_runtime.c:214]: Remove , mq, mc

File: src/vec_int512_runtime.c (hunk 5 above)
Comment (munroesj52) [reply src/vec_int512_runtime.c:249]: Removed mc,

File: src/vec_int512_runtime.c (hunk 6 above)
Comment (munroesj52) [reply src/vec_int512_runtime.c:280]: removed mc, mp, mq;

File: src/vec_int512_runtime.c (hunk 7 above)
Comment (munroesj52) [reply src/vec_int512_runtime.c:312]: Removed mc, mp, mq;

File: src/vec_int512_runtime.c (hunk 8 above)
Comment (munroesj52) [reply src/vec_int512_runtime.c:669]: OK

**Pull request merged**